File: Manuals/LJV_DllSampleAll_CPP/CPP/Source/LJV_DllSampleAll/ThreadSafeBuffer.cpp
//Copyright(c) 2013, KEYENCE Corporation. All rights reserved.
#include "StdAfx.h"
#include "ThreadSafeBuffer.h"

// Singleton Pattern

CThreadSafeBuffer* CThreadSafeBuffer::m_threadSafeBuffer = 0;

CThreadSafeBuffer::CThreadSafeBuffer(void)
{
	for(int i = 0; i < LJV7IF_DEVICE_COUNT; i++){ m_adwCount[i]  = 0; }
	for(int i = 0; i < LJV7IF_DEVICE_COUNT; i++){ m_adwNotify[i] = 0; }
	for(int i = 0; i < LJV7IF_DEVICE_COUNT; i++){ m_anBatchNo[i] = 0; }
}

CThreadSafeBuffer::~CThreadSafeBuffer(void)
{
}

CThreadSafeBuffer* CThreadSafeBuffer::getInstance(void)
{ 
	if(m_threadSafeBuffer == 0)
	{
		m_threadSafeBuffer = new CThreadSafeBuffer();
	}
	return m_threadSafeBuffer;
}

/*
 @note Add the profile data
 @param Current device ID
 @param Profile data
 @param notify
*/
void CThreadSafeBuffer::Add(DWORD dwIndex, vector<PROFILE_DATA> &vecProfileData, DWORD dwNotify)
{
	CSingleLock singleLock(&m_csDataAccess);
	singleLock.Lock();

	for(unsigned int i = 0; i < vecProfileData.size(); i++)
	{
		m_vecProfileData[dwIndex].push_back(vecProfileData.at(i));
	}

	m_adwCount[dwIndex]  += (unsigned int)vecProfileData.size();
	m_adwNotify[dwIndex] |= dwNotify; 
	if ((dwNotify & (0x1 << 16)) != 0) m_anBatchNo[dwIndex]++;
	
	singleLock.Unlock();
}

/*
 @note Add the profile count
 @param Current device ID
 @param Profile count
 @param notify
*/
void CThreadSafeBuffer::AddCount(DWORD dwIndex, DWORD dwCount, DWORD dwNotify)
{
	CSingleLock singleLock(&m_csDataAccess);
	singleLock.Lock();

	m_adwCount[dwIndex]  += dwCount;
	m_adwNotify[dwIndex] |= dwNotify; 
	if ((dwNotify & (0x1 << 16)) != 0) m_anBatchNo[dwIndex]++;
	
	singleLock.Unlock();
}

/*
 @note Get the profile
 @param Current device ID
 @param Notify
 @param Batch Number
 @param Profile data
*/
void CThreadSafeBuffer::Get(DWORD dwIndex, DWORD* pdwNotify, int* pnBatchNo, vector<PROFILE_DATA> &vecProfileData)
{
	CSingleLock singleLock(&m_csDataAccess);
	singleLock.Lock();

	vecProfileData = m_vecProfileData[dwIndex];		// Store the data
	m_vecProfileData[dwIndex].clear();				// Clear the data

	*pdwNotify = m_adwNotify[dwIndex];
	m_adwNotify[dwIndex] = 0;
	*pnBatchNo = m_anBatchNo[dwIndex];
	
	singleLock.Unlock();
}

/*
 @note Get the profile count
 @param Current device ID
 @param Notify
 @param Batch Number
*/
DWORD CThreadSafeBuffer::GetCount(DWORD dwIndex, DWORD* pdwNotify, int* pnBatchNo)
{
	CSingleLock singleLock(&m_csDataAccess);
	singleLock.Lock();

	*pdwNotify = m_adwNotify[dwIndex];
	m_adwNotify[dwIndex] = 0;
	*pnBatchNo = m_anBatchNo[dwIndex];

	return m_adwCount[dwIndex];
	
	singleLock.Unlock();
}

/*
 @note Clear the buffer data
 @param Current Device ID
*/
void CThreadSafeBuffer::ClearBuffer(int nCurrentDeviceID)
{
	CSingleLock singleLock(&m_csDataAccess);
	singleLock.Lock();

	m_vecProfileData[nCurrentDeviceID].clear();
	m_adwCount[nCurrentDeviceID]  = 0;
	m_anBatchNo[nCurrentDeviceID] = 0;
	m_adwNotify[nCurrentDeviceID] = 0;

	singleLock.Unlock();
}

File: Manuals/LJV_communication_library1_3/20130301_1_3_0_0 x86/ja/Sample/C#/Sample/Source/Sample/ThreadSafeBuffer.cs
//------------------------------------------------------------------------------ 
// <copyright file="ThreadSafeBuffer.cs" company="KEYENCE">
//     Copyright (c) 2012 KEYENCE CORPORATION.  All rights reserved.
// </copyright>
//----------------------------------------------------------------------------- 
using System.Collections.Generic;
using System;

namespace Sample
{
    /// <summary>
	/// 配列格納用スレッドセーフクラス
    /// </summary>
    public static class ThreadSafeBuffer
	{
		#region フィールド
		/// <summary>バッファ</summary>
		private static List<int[]> _buffer = new List<int[]>();
		/// <summary>排他制御用オブジェクト</summary>
		private static object syncObject = new object();
		/// <summary>通知フラグ</summary>
		private static uint _notify = 0;
		#endregion

		#region メソッド
		/// <summary>
		/// コンストラクタ
        /// </summary>
        static ThreadSafeBuffer()
        {
        }

		/// <summary>
		/// 要素の追加
        /// </summary>
		/// <param name="value">追加要素</param>
		/// <param name="notify">通知内容</param>
		public static void Add(List<int[]> value, uint notify)
        {
            lock (syncObject)
            {
				_buffer.AddRange(value);
				_notify |= notify;
            }
        }

        /// <summary>
		/// 要素のクリア
        /// </summary>
		public static void Clear()
        {
            lock (syncObject)
            {
				_buffer.Clear();
				_notify = 0;
            }
        }

        /// <summary>
		/// 要素の取得
        /// </summary>
		/// <param name="notify">通知内容</param>
		/// <returns>要素</returns>
		public static List<int[]> Get(ref uint notify)
        {
			List<int[]> value = new List<int[]>();
			lock (syncObject)
			{
				foreach(int[] data in _buffer)
				{
					value.Add(data);
				}
				_buffer.Clear();
				notify = _notify;
				_notify = 0;
			}
			return value;
        }

        #endregion
    }
}
